src/request_executor.cc

```diff
@@ -77,8 +77,8 @@ void
 InferResponseComplete(
     TRITONSERVER_InferenceResponse* response, const uint32_t flags, void* userp)
 {
-  auto linfer_payload = reinterpret_cast<std::shared_ptr<InferPayload>*>(userp);
-  std::unique_ptr<std::shared_ptr<InferPayload>> infer_payload(linfer_payload);
+  auto linfer_payload = reinterpret_cast<InferPayload*>(userp);
+  std::shared_ptr<InferPayload> infer_payload = linfer_payload->GetPtr();
```

Comment:
Is it valid to call `raw_ptr->GetPtr()` here?

I think it is because under the hood we know that `userp` was `infer_payload.get()`, where `infer_payload` was a `shared_ptr`, but I think it might a little sketchy/prone to future error. If this is indeed correct and don't have a clean alternative, I think we should document the constraints with a comment or two here.

> It is permitted to call shared_from_this only on a previously shared object, i.e. on an object managed by [std::shared_ptr](https://en.cppreference.com/w/cpp/memory/shared_ptr)<T>. Otherwise the behavior is undefined (until C++17)[std::bad_weak_ptr](https://en.cppreference.com/w/cpp/memory/bad_weak_ptr) is thrown (by the shared_ptr constructor from a default-constructed weak_this) (since C++17).
>
> enable_shared_from_this provides the safe alternative to an expression like [std::shared_ptr](https://en.cppreference.com/w/cpp/memory/shared_ptr)<T>(this), which is likely to result in this being destructed more than once by multiple owners that are unaware of each other (see example below).

ref:
- https://en.cppreference.com/w/cpp/memory/enable_shared_from_this
- https://stackoverflow.com/a/5548314

Reply:
Also can you summarize in words the current issue and what this PR is solving? 

The issue appears to be some lifetime issues passing the payload back and forth, and it looks like we're trying to solve this by passing `shared_ptr` through (`C++ <-> C API <-> C++`) to automatically manage the lifetime on both sides.

Reply:
Updated the PR description based on my understanding. @Tabrizian Please make any changes if needed.

Reply:
> Also can you summarize in words the current issue and what this PR is solving?
> 
> The issue appears to be some lifetime issues passing the payload back and forth, and it looks like we're trying to solve this by passing `shared_ptr` through (`C++ <-> C API <-> C++`) to automatically manage the lifetime on both sides.

Given the current design of the infer_payload being created as a shared pointer by python_be and then passed to request_executor, I believe this is cleanest approach. The basic problem was that the lifetime was managed by the shared pointer but the callback wasn't using a shared pointer and thus never increased the reference count.

That being said - I think if we were able to refactor to have the request_executor manage the creation and lifetime of infer_payload that may be cleaner (it doesn't seem at first glance that python_be needs to hang on to a reference - just the future). The creation and destruction ideally would be all handled internal to request_executor.

Reply:
> Is it valid to call `raw_ptr->GetPtr()` here?
> 
> I think it is because under the hood we know that `userp` was `infer_payload.get()`, where `infer_payload` was a `shared_ptr`, but I think it might a little sketchy/prone to future error if the caller changes the type from shared_ptr to something to unique or raw or something else in the future. If this is indeed correct and don't have a clean alternative, I think we should document the constraints with a comment or two here.
> 
> > It is permitted to call shared_from_this only on a previously shared object, i.e. on an object managed by [std::shared_ptr](https://en.cppreference.com/w/cpp/memory/shared_ptr). Otherwise the behavior is undefined (until C++17)[std::bad_weak_ptr](https://en.cppreference.com/w/cpp/memory/bad_weak_ptr) is thrown (by the shared_ptr constructor from a default-constructed weak_this) (since C++17).
> > enable_shared_from_this provides the safe alternative to an expression like [std::shared_ptr](https://en.cppreference.com/w/cpp/memory/shared_ptr)(this), which is likely to result in this being destructed more than once by multiple owners that are unaware of each other (see example below).
> 
> ref:
> 
> * https://en.cppreference.com/w/cpp/memory/enable_shared_from_this
> * https://stackoverflow.com/a/5548314

@Tabrizian also had an alternative that involved dynamically allocating a shared pointer which would be passed instead of the raw infer_payload. That method also maintains the lifetime of the object. In that case we would have a pointer to a shared pointer pointing to a infer_payload. For me that was more confusing - but open to reconsider.

Reply:
Thanks @krishung5 for updating the description. Looks accurate to me!